File: Day04/Part02/day04_02.py
import itertools


def get_anagrams(word):
    return ["".join(i) for i in itertools.permutations(word)]


if __name__ == '__main__':
    
    input_file = open("input.txt", "r")

    next_line = input_file.readline()

    valid_passphrase_count = 0

    while (next_line):

        passphrase = next_line.split()
        valid_passphrase = True
        used_words = []

        for word in passphrase:
            if word in used_words:
                valid_passphrase = False
            used_words.extend(get_anagrams(word))

        if valid_passphrase:
            valid_passphrase_count += 1

        next_line = input_file.readline()

    input_file.close()

    print("Answer:", valid_passphrase_count)

File: Day04/Part02/day04_02.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.util.ArrayList;
import java.util.HashSet;


public class day04_02 {

    public static void main(String[] args) {

        try {

            int validPassphraseCount = 0;

            BufferedReader br = new BufferedReader(new FileReader("input.txt"));

            String nextLine = br.readLine();

            while (nextLine != null) {
                Boolean validPassphrase = true;
                String[] splitLine = nextLine.split("\\W+");
                HashSet<String> usedWords = new HashSet<String>();
                for (String s: splitLine) {
                    // if the string is already added, then the passphrase is no good
                    if (!usedWords.add(s)) {
                        validPassphrase = false;
                    }
                    // add all possible combinations for next loop
                    for (String st: getCombinations(s)) {
                        usedWords.add(st);
                    }
                }
                if (validPassphrase) {
                    validPassphraseCount += 1;
                }
                nextLine = br.readLine();
            }

            System.out.println("Answer: " + validPassphraseCount);

            br.close();

        } catch (Exception e) {
            e.printStackTrace();
        }
        
    }

    public static String[] getCombinations(String word) {
        ArrayList<String> ans = new ArrayList<String>();
        if (word.length() <= 1) {
            ans.add(word);
        }
        else if (word.length() == 2) {
            ans.add(word);
            ans.add("" + word.charAt(1) + word.charAt(0));
        }
        else {
            for (int i = 0; i < word.length(); i++) {
                for (String s: getCombinations(word.substring(0, i) + word.substring(i+1))) {
                    ans.add(word.charAt(i) + s);
                }
            }
        }
        return ans.toArray(new String[ans.size()]);
    }

}
